libsolidity/formal/BMC.h

```diff
@@ -0,0 +1,171 @@
+	void internalOrExternalFunctionCall(FunctionCall const& _funCall);
+
+	/// Creates underflow/overflow verification targets.
+	std::pair<smt::Expression, smt::Expression> arithmeticOperation(
+		Token _op,
+		smt::Expression const& _left,
+		smt::Expression const& _right,
+		TypePointer const& _commonType,
+		Expression const& _expression
+	) override;
+
+	void resetStorageReferences();
+	void reset();
+
+	std::pair<std::vector<smt::Expression>, std::vector<std::string>> modelExpressions();
+	//@}
+
+	/// Verification targets.
+	//@{
+	struct VerificationTarget
+	{
+		enum class Type { ConstantCondition, Underflow, Overflow, UnderOverflow, DivByZero, Balance, Assert } type;
+		smt::Expression value;
+		smt::Expression constraints;
+		Expression const* expression;
+		std::vector<CallStackEntry> callStack;
+		std::pair<std::vector<smt::Expression>, std::vector<std::string>> modelExpressions;
+	};
+
+	void checkVerificationTargets(smt::Expression const& _constraints);
+	void checkVerificationTarget(VerificationTarget& _target, smt::Expression const& _constraints = smt::Expression(true));
+	void checkConstantCondition(VerificationTarget& _target);
+	void checkUnderflow(VerificationTarget& _target, smt::Expression const& _constraints);
+	void checkOverflow(VerificationTarget& _target, smt::Expression const& _constraints);
+	void checkDivByZero(VerificationTarget& _target);
+	void checkBalance(VerificationTarget& _target);
+	void checkAssert(VerificationTarget& _target);
+	void addVerificationTarget(
+		VerificationTarget::Type _type,
+		smt::Expression const& _value,
+		Expression const* _expression
+	);
+	//@}
+
+	/// Solver related.
+	//@{
+	/// Check that a condition can be satisfied.
+	void checkCondition(
+		smt::Expression _condition,
+		std::vector<CallStackEntry> const& callStack,
+		std::pair<std::vector<smt::Expression>, std::vector<std::string>> const& _modelExpressions,
+		langutil::SourceLocation const& _location,
+		std::string const& _description,
+		std::string const& _additionalValueName = "",
+		smt::Expression const* _additionalValue = nullptr
+	);
+	/// Checks that a boolean condition is not constant. Do not warn if the expression
+	/// is a literal constant.
+	/// @param _description the warning string, $VALUE will be replaced by the constant value.
+	void checkBooleanNotConstant(
+		Expression const& _condition,
+		smt::Expression const& _constraints,
+		smt::Expression const& _value,
+		std::vector<CallStackEntry> const& _callStack,
+		std::string const& _description
+	);
+	std::pair<smt::CheckResult, std::vector<std::string>>
+	checkSatisfiableAndGenerateModel(std::vector<smt::Expression> const& _expressionsToEvaluate);
+
+	smt::CheckResult checkSatisfiable();
+	//@}
+
+	/// Flags used for better warning messages.
+	bool m_loopExecutionHappened = false;
+	bool m_externalFunctionCallHappened = false;
+
+	/// ErrorReporter that comes from CompilerStack.
+	langutil::ErrorReporter& m_outerErrorReporter;
```

Comment:
Why outer? Isn't it always coming from somewhere else?

Reply:
As clarified in RL: the SMTEncoder keeps a local error reporter which is filled with the SMT warnings and transferred to the outer one (from CompilerStack) in case there were solvers available. If there are no solvers available it still creates the queries but does not report warnings.